bftengine/src/bftengine/SigManager.cpp

```diff
@@ -271,18 +251,103 @@ bool SigManager::verifySig(
         metrics_component_.UpdateAggregator();
     }
   }
+  LOG_DEBUG(GL,
+            "NON-replica validation result: " << KVLOG(
+                result, myId_, pid, sigLength, idOfReadOnlyReplica, idOfExternalClient));
   return result;
 }
 
-size_t SigManager::sign(const concord::Byte* data, size_t dataLength, concord::Byte* outSig) const {
-  return mySigner_->signBuffer(data, dataLength, outSig);
+size_t SigManager::sign(SeqNum seq, const concord::Byte* data, size_t dataLength, concord::Byte* outSig) const {
+  ConcordAssert(replicasInfo_.isIdOfReplica(myId_) || replicasInfo_.isRoReplica());
+  concord::crypto::ISigner* rawSigner = nullptr;
+  std::shared_ptr<EdDSAMultisigSigner> signer;
+  if (ReplicaConfig::instance().singleSignatureScheme) {
+    signer = CryptoManager::instance().getSigner(seq);
+    rawSigner = signer.get();
+  } else {
+    rawSigner = mySigner_.get();
+  }
+  auto result = rawSigner->signBuffer(data, dataLength, outSig);
+  LOG_DEBUG(GL, "Signing as replica with " << KVLOG(myId_, seq, rawSigner->signatureLength(), result));
+  return result;
 }
 
-size_t SigManager::sign(const char* data, size_t dataLength, char* outSig) const {
-  return sign(reinterpret_cast<const uint8_t*>(data), dataLength, reinterpret_cast<uint8_t*>(outSig));
+size_t SigManager::sign(SeqNum seq, const char* data, size_t dataLength, char* outSig) const {
+  return sign(seq, reinterpret_cast<const uint8_t*>(data), dataLength, reinterpret_cast<uint8_t*>(outSig));
+}
+
+bool SigManager::verifyReplicaSig(PrincipalId replicaID,
+                                  const concord::Byte* data,
+                                  size_t dataLength,
+                                  const concord::Byte* sig,
+                                  uint16_t sigLength) const {
+  LOG_DEBUG(GL, KVLOG(myId_));
+  ConcordAssert(replicasInfo_.isIdOfReplica(myId_) || replicasInfo_.isRoReplica());
+  ConcordAssert(replicasInfo_.isIdOfReplica(replicaID));
+  int i = 0;
+  for (auto [seq, multisigVerifier] : CryptoManager::instance().getLatestVerifiers()) {
```

Comment:
multisigVerifier -> verifier?
Why multisig?

Reply:
There were casts inside this function in a previous version.